File: Cpp/32_Longest_Valid_Parentheses.cpp
class Solution {
public:
    int longestValidParentheses(string s) {
        int result = 0;
        stack<int> st;

        for(int i = 0; i < s.size(); ++i) {
            if(s[i] == '(') {
                st.push(i);
            } else if(!st.empty() && s[st.top()] == '(') {
                st.pop();
                result = max(result, st.empty() ? i + 1 : i - st.top());
            } else {
                st.push(i);
            }
        }

        return result;
    }
};

File: Kotlin/32_Longest_Valid_Parentheses.kt
class Solution {
    fun longestValidParentheses(s: String): Int {
        var result = 0
        val st = Stack<Int>()

        for(i in s.indices) {
            if(s[i] == '(') {
                st.push(i)
            } else if(!st.isEmpty() && s[st.peek()] == '(') {
                st.pop()
                val current = if(st.isEmpty()) i + 1 else i - st.peek()
                result = maxOf(result, current)
            } else {
                st.push(i)
            }
        }

        return result
    }
}